contracts/finance/andromeda-weighted-distribution-splitter/src/contract.rs

```diff
@@ -0,0 +1,1145 @@
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    require(
+        !msg.recipients.is_empty(),
+        ContractError::EmptyRecipientsList {},
+    )?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let splitter = Splitter {
+        recipients: msg.recipients,
+        locked: false,
+    };
+
+    SPLITTER.save(deps.storage, &splitter)?;
+    ADOContract::default().instantiate(
+        deps.storage,
+        deps.api,
+        info,
+        BaseInstantiateMsg {
+            ado_type: "weighted-splitter".to_string(),
+            operators: None,
+            modules: msg.modules,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+
+    // Do this before the hooks get fired off to ensure that there is no conflict with the app
+    // contract not being whitelisted.
+    if let ExecuteMsg::AndrReceive(AndromedaMsg::UpdateAppContract { address }) = msg {
+        let splitter = SPLITTER.load(deps.storage)?;
+        let mut andr_addresses: Vec<AndrAddress> = vec![];
+        for recipient in splitter.recipients {
+            if let Recipient::ADO(ado_recipient) = recipient.recipient {
+                andr_addresses.push(ado_recipient.address);
+            }
+        }
+        return contract.execute_update_app_contract(deps, info, address, Some(andr_addresses));
+    };
+
+    contract.module_hook::<Response>(
+        deps.storage,
+        deps.api,
+        deps.querier,
+        AndromedaHook::OnExecute {
+            sender: info.sender.to_string(),
+            payload: encode_binary(&msg)?,
+        },
+    )?;
+
+    match msg {
+        ExecuteMsg::UpdateRecipients { recipients } => {
+            execute_update_recipients(deps, info, recipients)
+        }
+        ExecuteMsg::UpdateRecipientWeight { recipient } => {
+            execute_update_recipient_weight(deps, info, recipient)
+        }
+        ExecuteMsg::AddRecipient { recipient } => execute_add_recipient(deps, info, recipient),
+        ExecuteMsg::RemoveRecipient { recipient } => {
+            execute_remove_recipient(deps, info, recipient)
+        }
+        ExecuteMsg::UpdateLock { lock } => execute_update_lock(deps, info, lock),
+
+        ExecuteMsg::Send {} => execute_send(deps, info),
+        ExecuteMsg::AndrReceive(msg) => execute_andromeda(deps, env, info, msg),
+    }
+}
+
+pub fn execute_update_recipient_weight(
+    deps: DepsMut,
+    info: MessageInfo,
+    recipient: AddressWeight,
+) -> Result<Response, ContractError> {
+    // Only the contract's owner can update a recipient's weight
+    require(
+        ADOContract::default().is_contract_owner(deps.storage, info.sender.as_str())?,
```

Comment:
Might be best to make this owner or operator to allow other contracts to adjust weights if necessary